python/mim_estimation/ekf.py

```diff
@@ -1,243 +1,262 @@
-    def construct_continuous_noise_covariance(self, conf):
-        Qc = np.eye(self.__nx-3, dtype=float) #24x24
-        Qc[0:3, 0:3] = conf.Qf
-        Qc[3:6, 3:6] = conf.Qw
-        Qc[6:9, 6:9] = conf.Q_p1
-        Qc[9:12,9:12] = conf.Q_p2
-        Qc[12:15, 12:15] = conf.Q_p3
-        Qc[15:18, 15:18] = conf.Q_p4
-        Qc[18:21, 18:21] = conf.Q_bf
-        Qc[21:24, 21:24] = conf.Q_bw
+    def construct_continuous_noise_covariance(self):
+        Qc = np.zeros((self.__nx-3, self.__nx-3), dtype=float) #12x12
+        Qc[0:3, 0:3] = self.__Q_a
+        Qc[3:6, 3:6] = self.__Q_omega
+        Qc[6:9, 6:9] = self.__Qb_a
+        Qc[9:12,9:12] = self.__Qb_omega
         return Qc
     
     # using zero-order hold and truncating higher-order terms
     def construct_discrete_noise_covariance(self, Fk, Lc, Qc):
-        Qk_left = np.dot(np.dot(Fk, Lc), Qc)    
-        Qk_right = np.dot(Lc.T, Fk.T)
-        Qk = np.dot(Qk_left, Qk_right)*self.__dt
-        return Qk
-    
-    def construct_discrete_measurement_noise_covariance(self, conf):
-        R = (1/self.__dt)*conf.R
-        return R
-    
+        Qk_left = (Fk @ Lc)     
+        Qk_right = Lc.T @ Fk.T
+        return (Qk_left @ Qc @ Qk_right)*self.__dt
+   
+    def construct_discrete_measurement_noise_covariance(self):return (1/self.__dt)*self.__R
+
     # progagate covariance 
-    def prediction_step(self, Fk, Qk):
-        self.__Sigma = np.dot(np.dot(Fk, self.__Sigma), Fk.T) + Qk 
-    
-    def predict_measurements(self):
-        z = np.zeros(12, dtype=float)
-        C = self.get_base_orientation().rotation_matrix
-        z[0:3] = np.dot(C, (self.get_p1_position()-self.get_base_position()))
-        z[3:6] = np.dot(C, (self.get_p2_position()-self.get_base_position()))
-        z[6:9] = np.dot(C, (self.get_p3_position()-self.get_base_position()))
-        z[9:12] = np.dot(C, (self.get_p4_position()-self.get_base_position()))
-        return z
+    def prediction_step(self):
+        Fk = self.compute_discrete_prediction_jacobian()
+        Lc = self.compute_noise_jacobian()
+        Qc = self.construct_continuous_noise_covariance()
+        Qk = self.construct_discrete_noise_covariance(Fk, Lc, Qc)
+        self.__Sigma_pre = (Fk @ self.__Sigma_post @ Fk.T) + Qk 
     
-    def get_new_measurments(self, robot_configuration):
-        s = np.zeros(12, dtype=float)
-        end_effectors_positions = self.__compute_end_effectors_positions(robot_configuration)
-        C = self.__compute_base_pose_se3(robot_configuration).rotation_matrix
-        r = C.translation 
-        s[0:3] = np.dot(C, (end_effectors_positions['p1']-r))
-        s[3:6] = np.dot(C, (end_effectors_positions['p2']-r))
-        s[6:9] = np.dot(C, (end_effectors_positions['p3']-r))
-        s[9:12] = np.dot(C, (end_effectors_positions['p4']-r))
-        return s
+    # assuming contact logic variables coming from the contact schedule for now
+    # TODO contact logic prediction using a probabilistic model
```

Comment:
Using a probabalistic model for contact is a great idea (the Pronto paper includes this) but for now you may consider using the actual contact force assuming you can measure it reliably from joint torques? I had some issues just using the contact schedule from what I remember (on the humanoid though).

Reply:
Yes, thanks for raising this flag. I agree with this. I will check with @majidkhadiv this since he has done this previously with @Lhumd before.